File: Arrays_n_Strings/product_subarrays.py
# Find the number of subarrays that yield less than target product k

def num_of_subarray_product_less_than_k(arr, k):
    # Return 0 if the target k is less than or equal to 1
    if k <= 1:
        return 0

    left = 0
    current_product = 1
    num_of_subarrays = 0

    for right in range(len(arr)):
        current_product *= arr[right]

        # Shrink the window until the product is less than k
        while current_product >= k:
            current_product /= arr[left]
            left += 1

        # Update the count of subarrays
        num_of_subarrays += right - left + 1
    
    return num_of_subarrays

arr = [10, 5, 2, 6]
k = 100

print(num_of_subarray_product_less_than_k(arr, k))

File: Arrays_n_Strings/product_subarrays.js
/* Find the number of subarrays that yield less than target product k */

function numOfSubarrayProductLessThanK(arr, k) {

    // return no subarrays if the target k is less than
    // or equal to 1
    if (k <= 1) {
        return 0;
    }

    let left = 0;
    let currentProduct = 1;
    let numOfSubarrays = 0;

    for ( let right = 0; right < arr.length; right++ ) {
        currentProduct *= arr[right];

        while (currentProduct >= k) {
            currentProduct /= arr[left];
            left++;
        }

        numOfSubarrays += right - left + 1;
    }
    return numOfSubarrays;

}

let arr = [10, 5, 2, 6];
let k = 100;

console.log(numOfSubarrayProductLessThanK(arr, k))